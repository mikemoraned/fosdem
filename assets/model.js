import { createMergeableStore } from 'https://cdn.jsdelivr.net/npm/tinybase@5.4.4/+esm';
import { createBroadcastChannelSynchronizer } from 'https://cdn.jsdelivr.net/npm/tinybase@5.4.4/synchronizers/synchronizer-broadcast-channel/+esm';
import { createLocalPersister } from 'https://cdn.jsdelivr.net/npm/tinybase@5.4.4/persisters/persister-browser/+esm';

export async function createModel() {
    const store = createMergeableStore('fosdem2025');
    
    const persister = createLocalPersister(store, 'fosdem2025');
    await persister.load();
    await persister.startAutoSave()

    const synchronizer = createBroadcastChannelSynchronizer(
        store,
        'fosdem2025SyncChannel',
        () => {
            console.log('sent a message');
        },
        () => {
            console.log('received a message');
        }
      );
    synchronizer.addStatusListener((synchronizer, status) => {
        console.log(
          `${synchronizer.getChannelName()} channel status changed to ${status}`,
        );
      });
    await synchronizer.startSync();
    return new Model(store);
}

class Model {
    constructor(store) {
        this.store = store
    }

    getBookmarkStatus(eventId) {
        return this.store.getValue(eventId) || false;
    }

    setBookmarkStatus(eventId, status) {
        console.log(`Setting bookmark status for event ${eventId} to ${status}`);
        this.store.setValue(eventId, status);
        console.log(this.store.getContent());
        console.log(this.store.getMergeableContent());
    }

    addEventListener(eventId, listenerFn) {
        this.store.addValuesListener((store, getValueChange) => {
            const [hasChanged,_oldValue,newValue] = getValueChange(eventId);
            if (hasChanged) {
                console.log(`Event ${eventId} changed to ${newValue}`);
                listenerFn(newValue);
            }
        }, false);
    }

    exportEventIdsAsText() {
        const eventIds = [];
        this.store.forEachValue((eventId, isBookmarked) => {
            if (isBookmarked) {
                eventIds.push(eventId);
            }
        });
        return eventIds.join(' ');
    }

    importEventIdsFromText(text) {
        const possibleEventIds = text.split(' ');
        possibleEventIds.forEach((possibleEventId) => {
            if (possibleEventId.match(/\d+/)) {
                this.store.setValue(possibleEventId, true);
            }
        });
    }
}