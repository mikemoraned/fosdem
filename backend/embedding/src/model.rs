use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct SubjectEmbedding {
    subject: EventArtefact,
}

impl SubjectEmbedding {
    pub fn new(subject: EventArtefact) -> SubjectEmbedding {
        SubjectEmbedding { subject }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub enum EventArtefact {
    Combined { event_id: EventId },
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct EventId(pub u32);
