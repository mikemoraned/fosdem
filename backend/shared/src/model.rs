use chrono::{Duration, NaiveDate, NaiveDateTime, NaiveTime};
use nalgebra::DVector;
use openai_dive::v1::resources::embedding::EmbeddingResponse;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Debug, Clone, PartialEq)]
pub struct SearchItem {
    pub event: Event,
    pub distance: f64,
    pub related: Option<Vec<SearchItem>>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Event {
    pub id: u32,
    pub date: NaiveDate,
    pub start: NaiveTime,
    pub duration: u32,
    pub room: String,
    pub track: String,
    pub title: String,
    pub slug: String,
    pub url: Url,
    pub r#abstract: String,
    pub slides: Vec<Url>,
    pub presenters: Vec<Person>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Person {
    pub id: u32,
    pub name: String,
}

impl Event {
    pub fn starting_time(&self) -> NaiveDateTime {
        self.date.and_time(self.start)
    }

    pub fn ending_time(&self) -> NaiveDateTime {
        self.starting_time() + Duration::minutes(self.duration.into())
    }

    pub fn sojourner_url(&self) -> Url {
        let base_url = Url::parse("https://fosdem.sojourner.rocks/2024/event/").unwrap();
        base_url.join(&self.id.to_string()).unwrap()
    }

    pub fn nav_url(&self, current_event: &Option<Event>) -> Url {
        match current_event {
            Some(event) => Url::parse(&format!(
                "https://nav.fosdem.org/r/{}/{}",
                event.nav_room(),
                self.nav_room()
            ))
            .unwrap(),
            None => {
                let location_base_url = Url::parse("https://nav.fosdem.org/l/").unwrap();
                location_base_url.join(&self.nav_room()).unwrap()
            }
        }
    }

    pub fn nav_room(&self) -> String {
        if self.room.contains(' ') {
            let parts: Vec<_> = self.room.split(' ').collect();
            let start = parts[0];
            start.to_lowercase().replace('.', "")
        } else {
            self.room.to_lowercase().replace('.', "")
        }
    }
}

pub type OpenAIVector = DVector<f64>;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct OpenAIEmbedding {
    pub title: String,
    pub embedding: OpenAIVector,
}

impl OpenAIEmbedding {
    pub fn embedding_from_response(response: &EmbeddingResponse) -> OpenAIVector {
        let parts = response.data[0].embedding.clone();
        OpenAIVector::from(parts)
    }
}

#[derive(Debug, Clone)]
pub struct NextEvents {
    pub now: NaiveDateTime,
    pub current: Vec<Event>,
    pub selected: Event,
    pub next: Vec<Event>,
}

#[derive(Debug)]
pub enum NextEventsContext {
    Now,
    EventId(u32),
}
