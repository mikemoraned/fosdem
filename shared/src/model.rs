use std::fmt::{Display, Formatter};
use chrono::{Duration, NaiveDate, NaiveDateTime, NaiveTime};
use nalgebra::DVector;
use openai_dive::v1::resources::embedding::{EmbeddingOutput, EmbeddingResponse};
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Debug, Clone, PartialEq)]
pub struct SearchItem {
    pub event: Event,
    pub distance: f64,
    pub related: Option<Vec<SearchItem>>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Event {
    pub id: u32,
    pub guid: String,
    pub date: NaiveDate,
    pub start: NaiveTime,
    pub duration: u32,
    pub room: String,
    pub track: String,
    pub title: String,
    pub slug: String,
    pub url: Url,
    pub r#abstract: String,
    pub slides: Vec<Url>,
    pub presenters: Vec<Person>,
    pub links: Vec<Link>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Person {
    pub id: u32,
    pub name: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Link {
    pub url: Url,
    pub name: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct RoomId(String); // TODO: use RoomId instead of String in Event

impl RoomId {
    pub fn new(id: String) -> RoomId {
        RoomId(id)
    }

    pub fn nav_url(&self) -> Url {
        let location_base_url = Url::parse("https://nav.fosdem.org/l/").unwrap();
        location_base_url.join(&self.nav_room()).unwrap()
    }

    pub fn nav_room(&self) -> String {
        if self.0.contains(' ') {
            let parts: Vec<_> = self.0.split(' ').collect();
            let start = parts[0];
            start.to_lowercase().replace('.', "")
        } else {
            self.0.to_lowercase().replace('.', "")
        }
    }
}

impl Display for RoomId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl Event {
    pub fn starting_time(&self) -> NaiveDateTime {
        self.date.and_time(self.start)
    }

    pub fn ending_time(&self) -> NaiveDateTime {
        self.starting_time() + Duration::minutes(self.duration.into())
    }

    pub fn sojourner_url(&self) -> Url {
        let base_url = Url::parse("https://fosdem.sojourner.rocks/2025/event/").unwrap();
        base_url.join(&self.guid.to_string()).unwrap()
    }

    pub fn nav_url(&self, current_event: &Option<Event>) -> Url {
        match current_event {
            Some(event) => Url::parse(&format!(
                "https://nav.fosdem.org/r/{}/{}",
                event.nav_room(),
                self.nav_room()
            ))
            .unwrap(),
            None => {
                let location_base_url = Url::parse("https://nav.fosdem.org/l/").unwrap();
                location_base_url.join(&self.nav_room()).unwrap()
            }
        }
    }

    pub fn nav_room(&self) -> String {
        if self.room.contains(' ') {
            let parts: Vec<_> = self.room.split(' ').collect();
            let start = parts[0];
            start.to_lowercase().replace('.', "")
        } else {
            self.room.to_lowercase().replace('.', "")
        }
    }

    pub fn mp4_video_link(&self) -> Option<Url> {
        let video_links: Vec<Url> = self
            .links
            .iter()
            .filter(|l| l.name == "Video recording (mp4)" && l.url.to_string().ends_with(".mp4"))
            .map(|l| l.url.clone())
            .collect();
        video_links.first().cloned()
    }
}

pub type OpenAIVector = DVector<f64>;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct OpenAIEmbedding {
    pub title: String,
    pub embedding: OpenAIVector,
}

impl OpenAIEmbedding {
    pub fn embedding_from_response(response: &EmbeddingResponse) -> Result<OpenAIVector, Box<dyn std::error::Error>> {
        let output = response.data[0].embedding.clone();
        match output {
            EmbeddingOutput::Float(parts) => Ok(OpenAIVector::from(parts)),
            EmbeddingOutput::Base64(base64) => {
                Err(format!("Base64 encoding not supported: {}", base64).into())
            }
        }
    }
}

#[derive(Debug, Clone)]
pub struct NextEvents {
    pub now: NaiveDateTime,
    pub current: Vec<Event>,
    pub selected: Event,
    pub next: Vec<Event>,
}

#[derive(Debug)]
pub enum NextEventsContext {
    Now,
    EventId(u32),
}
