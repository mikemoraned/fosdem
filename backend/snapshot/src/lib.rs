use shared::model::SearchItem;

pub struct Snapshotter;

impl Snapshotter {
    pub fn new() -> Snapshotter {
        Snapshotter {}
    }

    pub fn find_similar(&self, _phrase: &str) -> Vec<SearchItem> {
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_phrase() {
        let snapshotter = Snapshotter::new();
        let similar = snapshotter.find_similar("controversial");
        insta::assert_yaml_snapshot!(similar);
    }
}
