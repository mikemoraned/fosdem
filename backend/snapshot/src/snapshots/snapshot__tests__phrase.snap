---
source: snapshot/src/lib.rs
expression: similar
---
[]
